onnxruntime/contrib_ops/cuda/bert/group_query_attention_impl.cu

```diff
@@ -139,129 +142,475 @@
-        ConcatNewToPastKVLarge<float2><<<grid, block, 0, stream>>>(kv_sequence_length,
-                                                                   H,
-                                                                   reinterpret_cast<const float2*>(data.past_key),
-                                                                   reinterpret_cast<const float2*>(data.key),
-                                                                   reinterpret_cast<float2*>(data.present_key),
-                                                                   past_kv_format == AttentionQkvFormat::Q_K_V_BSNH);
-        ConcatNewToPastKVLarge<float2><<<grid, block, 0, stream>>>(kv_sequence_length,
-                                                                   H,
-                                                                   reinterpret_cast<const float2*>(data.past_value),
-                                                                   reinterpret_cast<const float2*>(data.value),
-                                                                   reinterpret_cast<float2*>(data.present_value),
-                                                                   past_kv_format == AttentionQkvFormat::Q_K_V_BSNH);
-      }
-
-      void* present_key = reinterpret_cast<void*>(const_cast<T*>(data.present_key));
-      void* present_value = reinterpret_cast<void*>(const_cast<T*>(data.present_value));
-
-      // Launch kernel to copy seqlen
-      int thr_per_blk = 256;
-      int blk_in_grid = ceil(float(batch_size) / thr_per_blk);
-      repeat_seqlen<<<blk_in_grid, thr_per_blk, 0, stream>>>(data.seqlens_k, parameters.past_sequence_length, batch_size);
-
-      bool past_bsnh = past_kv_format == AttentionQkvFormat::Q_K_V_BSNH;
-      ORT_RETURN_IF_ERROR(onnxruntime::flash::mha_fwd(
-          device_prop, stream, query, present_key, present_value, data.output, reinterpret_cast<void*>(data.softmax_lse),
-          batch_size, num_heads, kv_num_heads, head_size,
-          sequence_length, present_sequence_length, scale, is_causal, parameters.num_splits,
-          reinterpret_cast<void*>(data.softmax_lse_accum), reinterpret_cast<void*>(data.out_accum), past_bsnh));
+  const void* query = reinterpret_cast<const void*>(data.query);
+  const void* key = reinterpret_cast<const void*>(data.key);
+  const void* value = reinterpret_cast<const void*>(data.value);
+  int final_kv_seqlen = kv_sequence_length;
+  if (data.present_key != nullptr) {
+    // concatenate new kv to past kv
+    if (data.past_key == data.present_key) {
+      ORT_RETURN_IF_ERROR(LaunchConcatKVInPlace(parameters, data, stream, max_threads_per_block));
+      final_kv_seqlen = past_sequence_length + kv_sequence_length;
+    } else {
+      ORT_RETURN_IF_ERROR(LaunchConcatNewToPastKV(parameters, data, stream, max_threads_per_block));
+      final_kv_seqlen = present_sequence_length;
+    }
+    const bool is_bsnh = past_kv_format == AttentionQkvFormat::Q_K_V_BSNH;
```

Comment:
>is_bsnh [](http://example.com/codeflow?start=15&length=7)

it is very expensive to transpose KV cache each time. We need to log a warning here to remind user to use BSNH KV cache format.

Reply:
MemoryEfficientAttention also supports BNSH K and V by setting the stride. We should set the stride instead of transpose the K and V just like what you did for FlashAttention: [https://github.com/microsoft/onnxruntime/blob/009cd4ea2e0621459806010cea7d7533d0acb39d/onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/fmha_launch_template.h#L58C5-L71](https://github.com/microsoft/onnxruntime/blob/009cd4ea2e0621459806010cea7d7533d0acb39d/onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/fmha_launch_template.h#L58C5-L71).